util/setuidgid.c

```diff
@@ -0,0 +1,332 @@
+	pos = strcspn(combined, ":");
+	len = strlen(combined);
+
+	/* Reject silly strings. */
+	if (pos == len - 1) {
+		warn0("Groupname is blank.");
+		goto err0;
+	}
+
+	/* String ok, proceed. */
+	if (pos == 0) {
+		/* Groupname only; duplicate string. */
+		if ((*groupname_p = strdup(&combined[1])) == NULL) {
+			warnp("strdup()");
+			goto err0;
+		}
+		*username_p = NULL;
+	} else if (pos != len) {
+		/* Extract username. */
+		if ((*username_p = malloc(pos + 1)) == NULL) {
+			warnp("Failed to allocate memory");
+			goto err0;
+		}
+		strncpy(*username_p, combined, pos);
+		(*username_p)[pos] = '\0';
+
+		/* Extract groupname. */
+		if ((*groupname_p = strdup(&combined[pos + 1])) == NULL) {
+			warnp("strdup()");
+			goto err1;
+		}
+	} else {
+		/* Duplicate string. */
+		if ((*username_p = strdup(combined)) == NULL) {
+			warnp("strdup()");
+			goto err0;
+		}
+		*groupname_p = NULL;
+	}
+
+	/* Success! */
+	return (0);
+
+err1:
+	free(*username_p);
+err0:
+	/* Failure! */
+	return (-1);
+}
+
+/**
+ * setuidgid(user_group_string, leave_suppgrp):
+ * Set the UID and/or GID to the names given in ${user_group_string}.
+ * Depending on the existence and position of a colon ":", the behaviour is
+ * - no ":" means that the string is a username.
+ * - ":" in the first position means that the string is a groupname.
+ * - otherwise, the string is parsed into "username:groupname".
+ *
+ * The behaviour with supplementary groups depends on ${leave_suppgrp}:
+ * - SETUIDGID_SGROUP_IGNORE: do not attempt to leave supplementary groups.
+ * - SETUIDGID_SGROUP_LEAVE_WARN: attempt to leave; if it fails, give a
+ *   warning but continue.
+ * - SETUIDGID_SGROUP_LEAVE_ERROR: attempt to leave; if it fails, return
+ *   an error.
+ */
+int
+setuidgid(const char * opt_usergroup, int leave_suppgrp)
+{
+	char * username;
+	char * groupname;
+	int eperm = 0;
+
+	/* Get the username:groupname from the opt_user string. */
+	if (string_extract_user_group(opt_usergroup, &username, &groupname))
+		goto err0;
+
+	/* If requested, leave supplementary groups. */
+	if (leave_suppgrp != SETUIDGID_SGROUP_IGNORE) {
+		/* Leave any supplementary groups. */
+		if (setgroups_none()) {
+			if (leave_suppgrp == SETUIDGID_SGROUP_LEAVE_ERROR)
+				goto err1;
+		}
+
+		/* Check that we have actually left all supplementary groups. */
+		if (check_supplementary_groups_none()) {
+			if (leave_suppgrp == SETUIDGID_SGROUP_LEAVE_ERROR) {
+				warn0("Failed to leave supplementary groups");
+				goto err1;
+			} else {
+				warn0("Warning: Failed to leave supplementary "
+				    "groups");
+			}
+		}
+	}
+
+	/* Set group; must be changed before user. */
+	if (groupname && set_group(groupname))
+		goto err1;
+
+	/* Set user. */
+	if (username && set_user(username))
+		goto err1;
+
+	/* Clean up. */
+	free(groupname);
+	free(username);
+
+	/* Success! */
+	return (0);
+
+err1:
+	/* POSIX Issue 7 does not forbid free() from modifying errno. */
+	if (errno == EPERM)
```

Comment:
`saved_errno = errno;` is more idiomatic... is there a reason we only want to save EPERM and don't want to preserve any other error code from `free`'s intervention?

Reply:
oh, also: every error path other than `EPERM` should have already received a `warnp()` or `warn0()`.
(I'll double-check this tomorrow)